src/hittable.rs

```diff
@@ -34,41 +32,78 @@ impl<'a> Hit<'a> {
     }
 }
 
-
-pub trait Hittable {
+pub trait Hittable: Debug {
     fn hit(&self, ray: &Ray, dist_min: f64, dist_max: f64) -> Option<Hit>;
+    fn bounding_box(&self, t_min: f32, t_max: f32) -> Option<AABB> { None }
 }
 
 
+#[derive(Debug)]
 pub struct Sphere {
     center: V3,
     radius: f64,
+    aabb: AABB,
     material: Box<dyn Material>,
 }
 
-trait SphereT {
-    fn center(&self, time: f32) -> V3;
-    fn radius(&self) -> f64;
-    fn material<'a>(&'a self) -> &'a Box<dyn Material>;
-}
-
-
 impl Sphere {
     pub fn new(center: V3, radius: f64, material: Box<dyn Material>) -> Sphere {
-        Sphere { center, radius, material }
+        let aabb = AABB::new(center - radius, center + radius);
+        Sphere { center, radius, aabb, material }
+    }
+    fn center(&self, time: f32) -> V3 {
+        self.center
+    }
+    fn radius(&self) -> f64 { self.radius }
+    fn material<'a>(&'a self) -> &'a Box<dyn Material> { &self.material }
+
+    fn aabb(&self, t0: f32, t1: f32) -> AABB {
+        AABB::new(self.center - self.radius, self.center + self.radius)
     }
 }
 
-impl SphereT for Sphere {
+#[derive(Debug)]
+pub struct MovingSphere {
+    center_t0: V3,
+    center_t1: V3,
+    time0: f32,
+    duration: f32,
+    aabb: AABB,
+    radius: f64,
+    material: Box<dyn Material>,
+}
+
+impl MovingSphere {
+    pub fn new(center_t0: V3, center_t1: V3, time0: f32, time1: f32, radius: f64, material: Box<dyn Material>) -> MovingSphere {
+        let aabb0 = AABB::new(center_t0 - radius, center_t0 + radius);
+        let aabb1 = AABB::new(center_t1 - radius, center_t1 + radius);
+        MovingSphere {
+            center_t0,
+            center_t1,
+            time0,
+            duration: time1 - time0,
+            aabb: aabb0 + aabb1,
+            radius,
+            material,
+        }
+    }
     fn center(&self, time: f32) -> V3 {
-        self.center
+        let scale = (time - self.time0) / self.duration;
+        self.center_t0 + scale * (self.center_t1 - self.center_t0)
```

Comment:
[`f64::mul_add`](https://doc.rust-lang.org/std/primitive.f64.html#method.mul_add) can be used here for a more accurate result